infrastructure/src/main/java/org/corfudb/infrastructure/redundancy/RedundancyCalculator.java

```diff
@@ -222,6 +223,38 @@ public static boolean canRestoreRedundancyOrMergeSegments(Layout layout, String
                 .collect(ImmutableList.toImmutableList());
     }
 
+    /**
+     * Get all the layout segments that were trimmed and that do not contain a current server, and
+     * return them as TRANSFERRED segments. Since these segments were trimmed, they
+     * contain zero transferable addresses, and hence are not eligible for state transfer.
+     * However, since they also do not contain a current server, they should be considered
+     * TRANSFERRED, eligible for the layout redundancy restoration.
+     *
+     * @param layout   A current layout.
+     * @param trimMark A current global trim mark.
+     * @return A list of transfer segments.
+     */
+    public ImmutableList<TransferSegment> getTrimmedNotRestoredSegments(Layout layout, long trimMark) {
+        return layout.getSegments()
+                .stream()
+                .filter(segment -> segment.getEnd() <= trimMark && !segmentContainsServer(segment, getServer()))
```

Comment:
maybe use two filter methods?
```
.filter(segment -> segment.getEnd() <= trimMark)
.filter(segment ->  !segmentContainsServer(segment, getServer()))
```

Reply:
Usually, when there are multiple consecutive HOFs I think it's better to merge them into one HOF. For example,
```
.map(x -> x + 1)
.map(x -> x * 2) 
```
I think is better expressed as `.map(x -> (x + 1) * 2)`